Add redshift database plugin

malnick opened the pull request:
Adds a database secrets engine for Redshift (postgres 8 compatible). 

Because there are no local mocking methods available for Redshift, this plugin leverages a live Redshift cluster and enables test methods using the `VAULT_ACC` pattern. See comments in header to `redshift_test.go` for a full disclosure of caveats and setup instructions to run these acceptance tests. 

Running these tests locally:
```
➜  vault/plugins/database/redshift malnick/redshift ✓ VAULT_ACC=1 REDSHIFT_URL=<redacted>:5439/<redacted> REDSHIFT_USER=<redacted> REDSHIFT_PASSWORD=<redacted> TEST_ROTATE_ROOT=1 go test -v
=== RUN   TestPostgreSQL_Initialize
--- PASS: TestPostgreSQL_Initialize (1.41s)
=== RUN   TestPostgreSQL_CreateUser
--- PASS: TestPostgreSQL_CreateUser (5.78s)
=== RUN   TestPostgreSQL_RenewUser
--- PASS: TestPostgreSQL_RenewUser (10.36s)
=== RUN   TestPostgreSQL_RevokeUser
--- PASS: TestPostgreSQL_RevokeUser (8.76s)
=== RUN   TestPostgresSQL_SetCredentials
--- PASS: TestPostgresSQL_SetCredentials (4.82s)
=== RUN   TestPostgreSQL_RotateRootCredentials
rotated root credentials, new user/pass:
username: <redacted>
password: <redacted>
--- PASS: TestPostgreSQL_RotateRootCredentials (1.29s)
PASS
ok      github.com/hashicorp/vault/plugins/database/redshift    33.138s
```

calvn commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,528 @@
+		return nil, errors.New("username and password are required to rotate")
+	}
+
+	rotateStatents := statements
```
Looks like the var is misspelled, `s/rotateStatents/rotateStatements/`.

malnick replied:
A transient artifact no less https://github.com/hashicorp/vault/blob/master/plugins/database/postgresql/postgresql.go#L452

I'll fix this and PR a fix in the original codebase too. Good catch, thanks!

malnick replied:
https://github.com/hashicorp/vault/pull/8306

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+		return "", "", err
+	}
+	defer func() {
+		_ = tx.Rollback()
```
```suggestion
		tx.Rollback()
```
Might as well just quietly drop the error, we do everywhere else. It would be super nice if we could log these, but alas, no logger!

tyrannosaurus-becks replied:
Also, I have a question about these Rollback statements. I see Rollback is called this way in the `mysql` plugin. Is the idea basically that if we succeeded doing the transaction, then Rollback just won't work? I'm wondering if maybe we should add a guard that's like "success" that is false by default, and then it gets set to true if the whole transaction succeeded. Then inside the deferred function, it would be like "if !success, rollback" (pseudocode).

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+ALTER USER "{{username}}" WITH PASSWORD '{{password}}';
+`
+
+	defaultRotateCredentialsSQL = `
```
This appears to be unused.

malnick replied:
removed

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+	return sqlTypeName, nil
+}
+
+func (p *RedShift) getConnection(ctx context.Context) (*sql.DB, error) {
```
Might be good to add a comment to this method that it's up to the caller to call `Close` on the returned obj. Very easy to not know about.

malnick replied:
fixed

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+	if err != nil {
+		return "", "", err
+	}
+
```
```suggestion
    defer db.Close()
```

malnick replied:
fixed

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+	if err != nil {
+		return "", "", err
+	}
+
```
```suggestion
    defer db.Close()
```

malnick replied:
fixed

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+	if err != nil {
+		return err
+	}
+
```
```suggestion
    defer db.Close()
```

malnick replied:
fixed

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+	if err != nil {
+		return err
+	}
+
```
```suggestion
    defer db.Close()
```

malnick replied:
fixed

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+	if err != nil {
+		return err
+	}
+
```
```suggestion
    defer db.Close()
```

malnick replied:
Fixed

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+	}
+
+	// Close the database connection to ensure no new connections come in
+	if err := db.Close(); err != nil {
```
Could we move this up to living in a defer statement on line 476? That way there's no way we can accidentally leave the connection open.

malnick replied:
Fixed

tyrannosaurus-becks commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+	var lastStmtError error
+	for _, query := range revocationStmts {
+		if err := dbtxn.ExecuteDBQuery(ctx, db, nil, query); err != nil {
+			lastStmtError = err
```
Might be nice to use [go-multierror](https://github.com/hashicorp/go-multierror) here to return more information about what did and didn't succeed. Commonly used in Vault.

tyrannosaurus-becks replied:
I'm also wondering about an edge case here. I do believe that revocations get retried if they return an error. So, if we got part way through here, returned an error below, and then a minute later restarted the logic from the top, could we get all the way down? I'm not sure we need to do somersaults to make it work - I notice the other plugins use simplistic logic, just something to think about.

malnick replied:
Do you have a good example of how this is implemented elsewhere? This plugin was largely inspired (and ripped off from) the PG secrets engine. If there's a simpler implementation you had in mind, would love to check it out.

tyrannosaurus-becks replied:
Sure, I think this is a nice example of it: https://github.com/hashicorp/vault/blob/master/plugins/database/influxdb/influxdb.go#L192

tyrannosaurus-becks replied:
You can also treat it like a regular error. So, for instance, in the example above, if right after the loop you wanted to do something like:
```
if result != nil {
    return
}
... do more stuff
```
It would totally work.

malnick replied:
Fixed

tyrannosaurus-becks commented on plugins/database/redshift/redshift_test.go:
```diff
@@ -0,0 +1,524 @@
+password on your Redshift cluster to an insecure, cleartext password defined in the
+test method. Because of this, you must pass TEST_ROTATE_ROOT=1 to enable it explicitly.
+
+Do not run this test suite against a production Redshift cluster.
```
:smile:

tyrannosaurus-becks commented on plugins/database/redshift/redshift_test.go:
```diff
@@ -0,0 +1,524 @@
+)
+
+/*
+To run these sets of acceptance tests, you must pre-configure a Redshift cluster
```
Just curious, since Redshift is similar to Postgres, do you think these tests might pass against a Postgres Docker image? Doesn't have to be something we do in this PR, could be a task during the testing cycle before the beta.

malnick replied:
I would love it if they did, I spent a day trying to make that work. Unfortunately, PG 8 is substantially different than modern postgres versions. Making this even more difficult is that Redshift has it's own nuances in SQL that conflict with the postgres 8 variant it's built on, making even postgres 8 docker images unusable in mocking Redshift. 

I did try with a PG 8 image FWIW, this testing decision was the last resort option.

tyrannosaurus-becks replied:
Well, it was a beautiful thought!

malnick commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+var _ dbplugin.Database = &RedShift{}
+
+// New implements builtinplugins.BuiltinFactory
+func New(lowercaseUsername bool) func() (interface{}, error) {
```
Great question! `lowercaseUsername` is actually the underlying cause of writing this entire plugin. Redshift implements (mostly) a postgres 8 interface (which is quite old), and part of that is under the hood, it's lowercasing the usernames (the exact opposite of the behavior we saw in Oracle, interestingly). 

Initially, @briankassouf thought we could get away with adding this to our PG client code, shared between many of our secrets engines. In the end, it required us to do that, plus build a specific codebase capable of handling the PG 8 requirements (table names and some SQL verbs varied slightly) of Redshift.

tyrannosaurus-becks replied:
Sounds good. It might be useful to add a comment here to that effect so people understand the history and reasoning. I do like the explicitness of keeping the variable, as well.

malnick commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+)
+
+const (
+	sqlTypeName     = "postgres"
```
I considered changing this to the suggestion initially, but it doesn't buy us to much:
1. Redshift is PG under the hood, so it still makes sense to re-use that connection logic
2. The only modification that the client code required was forcing timezone to uppercase (`utc`->`UTC`), and this change is acceptable for modern versions of PG (the integration tests on the PG plugin passed)
3. If #2 changes, then I think it makes sense to build an independent `redshift` variant of our PG client code

tyrannosaurus-becks replied:
Gotcha. I actually didn't see it was in use for the driver, I'd only noticed that it's returned as the database `Type`. :-) 

I totally agree about keeping it as the driver type. On the other use in the `Type` function, the name returned is used in metrics and other middleware. I'm just thinking that folks might be confused if they're thinking of it as a redshift plugin but all the outward references to it call it postgres.

malnick commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+`
+)
+
+var _ dbplugin.Database = &RedShift{}
```
Looks unused to me, good catch. I'll drop it.

malnick replied:
Fixed

malnick commented:
Thanks @tyrannosaurus-becks - I got to everything but the final question you had in here. I force pushed over the existing commit, here's the results of testing with the latest changes:
```
=== RUN   TestPostgreSQL_Initialize
--- PASS: TestPostgreSQL_Initialize (1.53s)
=== RUN   TestPostgreSQL_CreateUser
--- PASS: TestPostgreSQL_CreateUser (6.12s)
=== RUN   TestPostgreSQL_RenewUser
--- PASS: TestPostgreSQL_RenewUser (11.38s)
=== RUN   TestPostgreSQL_RevokeUser
--- PASS: TestPostgreSQL_RevokeUser (9.11s)
=== RUN   TestPostgresSQL_SetCredentials
--- PASS: TestPostgresSQL_SetCredentials (5.28s)
=== RUN   TestPostgreSQL_RotateRootCredentials
--- SKIP: TestPostgreSQL_RotateRootCredentials (0.00s)
PASS
ok      github.com/hashicorp/vault/plugins/database/redshift    33.818s```

malnick commented on plugins/database/redshift/redshift.go:
```diff
@@ -0,0 +1,519 @@
+	}
+}
+
+func new(lowercaseUsername bool) *RedShift {
```
Fixed

tyrannosaurus-becks commented:
Thanks for all your work on this so far! Would you be willing to merge in master so we can also get a clean look at the build before merging?

malnick commented:
Awesome, thanks @tyrannosaurus-becks - added the plugin to the plugin predict test file and rebased on master.

malnick commented:
Thanks for the second pass @tyrannosaurus-becks ! 

I added the missing `db.Close()` statements and fixed the remaining feedback. Rebased on master just now, but it'll probably need to happen one more time soon because we're merging code quickly right now.

Latest test results with fixes and rebase:
```
=== RUN   TestPostgreSQL_Initialize
--- PASS: TestPostgreSQL_Initialize (1.46s)
=== RUN   TestPostgreSQL_CreateUser
--- PASS: TestPostgreSQL_CreateUser (7.50s)
=== RUN   TestPostgreSQL_RenewUser
--- PASS: TestPostgreSQL_RenewUser (11.83s)
=== RUN   TestPostgreSQL_RevokeUser
--- PASS: TestPostgreSQL_RevokeUser (10.41s)
=== RUN   TestPostgresSQL_SetCredentials
--- PASS: TestPostgresSQL_SetCredentials (5.29s)
=== RUN   TestPostgreSQL_RotateRootCredentials
--- SKIP: TestPostgreSQL_RotateRootCredentials (0.00s)
PASS
ok      github.com/hashicorp/vault/plugins/database/redshift    38.058s```
